niceman/distributions/debian.py

```diff
@@ -368,93 +365,212 @@ def _find_all_sources(self):
+                            p.get("name"))
+                continue
+            p["source_name"] = r.get("Source_name")
+            p["source_version"] = r.get("Source_version")
+            p["size"] = r.get("Size")
+            p["md5"] = r.get("MD5sum")
+            p["sha1"] = r.get("SHA1")
+            p["sha256"] = r.get("SHA256")
+
+    def _get_pkgs_install_date(self, pkg_dicts):
+        # Convert package names to dpkg list filenames
+        queries = []
+        for p in pkg_dicts:
+            queries.append(self._pkg_name_to_dpkg_list_file(p.get("name")))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = {}
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
+            try:
+                out, _ = self._session.execute_command(
+                    ['stat', '-c', '%n: %Y'] + subqueries
+                )
+            except CommandError as exc:  # file not found
+                stderr = utils.to_unicode(exc.stderr, "utf-8")
+                if 'No such file or directory' in stderr:
+                    out = exc.stdout  # One file not found, so continue
+                else:
+                    raise  # some other fault -- handle it above
+            # Parse the output and store by filename
+            for outlines in out.splitlines():
+                (fname, ftime) = outlines.split(": ")
+                cmd_results[fname] = str(
+                    pytz.utc.localize(
+                        datetime.utcfromtimestamp(float(ftime))))
+        # Now lookup the packages in the results
+        for p in pkg_dicts:
+            fname = self._pkg_name_to_dpkg_list_file(p.get("name"))
+            if fname in cmd_results:
+                p["install_date"] = cmd_results[fname]
+
+    def _pkg_name_to_dpkg_list_file(self, name):
+        query = "/var/lib/dpkg/info/" + name + ".list"
+        return query
+
+    def _get_pkgs_versions_and_sources(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            queries.append(name if not architecture
+                           else "%s:%s" % (name, architecture))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = {}
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
```

Comment:
oy, it is haunting me!

Reply:
Now to see if my cure is worse than the original ailment...